src/controller/update.ts

```diff
@@ -0,0 +1,235 @@
+/*
+This file contains functions that run on each frame.
+*/
+/*
+global
+G_utils_getSign
+G_model_isKeyDown
+G_model_actorSetVelocity
+G_model_actorSetPosition
+G_model_actorSetAcceleration
+G_model_actorSetFacing
+G_model_actorSetAnimState
+G_model_roomGetSizePx
+G_model_roomGetCollidableTiles
+G_model_tileIsFullyCollidable
+G_model_getFrameMultiplier
+G_utils_floorNearestMultiple
+G_utils_createRect
+G_utils_getCollisionsWithRect
+G_utils_rectToCollisionPoints
+G_KEY_LEFT
+G_KEY_RIGHT
+G_KEY_SPACE
+G_FACING_LEFT
+G_FACING_RIGHT
+G_ANIM_JUMPING
+G_ANIM_WALKING
+G_ANIM_DEFAULT
+G_COLLISION_TOP
+G_COLLISION_BOTTOM
+G_COLLISION_LEFT
+G_COLLISION_RIGHT
+*/
+
+const PLAYER_SPEED = 1.2; // the acceleration rate that the player moves left/right
+const PLAYER_JUMP_SPEED = 5.1; // the higher this value, the higher+faster the player jumps
+const AIR_CONTROL_DIVISOR = 30; // the higher this value, the less air control the player has.
+const MAX_SPEED_X = 1.2; // the maximum x speed that an actor can move
+const MAX_SPEED_Y = 3.2; // the maximum y speed that an actor can move
+const DECELERATION_RATE = 1.2; // the rate that actors decelerate (reduce velocity) per frame
+const GRAVITY_RATE = 0.2; // the rate that the y velocity is modified each frame (the higher, the faster the player moves down)
+
+// given an actor and a 'vx' or 'vy' reduce the velocity value one step closer to 0
+const decelerateVelocity = (actor: Actor, vel: 'vx' | 'vy') => {
+  const v = actor[vel];
+
+  if (v < 0) {
+    actor[vel] += DECELERATION_RATE;
+  } else if (v > 0) {
+    actor[vel] -= DECELERATION_RATE;
+  }
+
+  if (Math.abs(actor[vel]) <= DECELERATION_RATE) {
+    actor[vel] = 0;
+  }
+};
+
+const applyGravity = (actor: Actor) => {
+  const fm = G_model_getFrameMultiplier();
+  if (actor.isGround) {
+    actor.ay += 0.0001 * fm;
+  } else {
+    actor.ay += GRAVITY_RATE * fm;
+  }
+};
+
+interface CollisionResult {
+  sides: CollisionSide[];
+  tile: Tile;
+}
+
+// check collisions between this actor and all wall tiles.  If the actor is colliding,
+// then move the actor towards a position that is not colliding until the actor is no
+// longer colliding
+const handleActorCollisions = (actor: Actor, room: Room) => {
+  let hasCollision: boolean;
+  let hasCollisionWithGround = false;
+  let ctr = 0;
+  const collidableTiles = G_model_roomGetCollidableTiles(room);
+  do {
+    ctr++;
+    hasCollision = false;
+    const actorCollisionPoints = G_utils_rectToCollisionPoints(
+      G_utils_createRect(actor.x, actor.y, actor.w, actor.h)
+    );
+
+    const collisionSideMap = {
+      [G_COLLISION_BOTTOM]: null,
+      [G_COLLISION_TOP]: null,
+      [G_COLLISION_LEFT]: null,
+      [G_COLLISION_RIGHT]: null,
+    } as any;
```

Comment:
I'm trying to save space.  A normal while loop requires you to declare all your variables for the first iteration outside the loop, but that step can be skipped by checking the condition last instead of first, and saves a small amount of code space.